Enhancement: Guided Tour update orientation

lovaiible opened the pull request:
### Checklist
- [x] I tested *all* changes and *all* related features with different users (student, tutor, instructor, admin) on the test server https://artemistest.ase.in.tum.de.
- [x] Client: I added multiple integration tests (Jest) related to the features
- [x] ~~Client: I added `authorities` to all new routes and check the course groups for displaying navigation elements (links, buttons)~~
- [x] Client: I documented the TypeScript code using JSDoc style.
- [x] Client: I added multiple screenshots/screencasts of my UI changes
- [x] ~~Client: I translated all the newly inserted strings into German and English~~

### Motivation and Context
This PR should solve the issue https://github.com/ls1intum/Artemis/issues/972
(Description in issue)

### Steps for Testing
1. All tutorials should work as usual
2. On the course overview page, the tour step for the `Tutorial Course` might be cut off depending on the position of the course tile. In that case the position of the tour step should be flipped (see screencast).

### Screenshots
![video](https://user-images.githubusercontent.com/33764166/72668173-3e7f8680-3a24-11ea-8b20-2203fccb7074.gif)

maxr96 commented on src/main/webapp/app/guided-tour/guided-tour.component.ts:
```diff
@@ -173,68 +178,191 @@ export class GuidedTourComponent implements AfterViewInit, OnDestroy {
+     */
+    private flipOrientation(): void {
+        if (this.isLeft()) {
+            switch (this.orientation) {
```
Hm here you are accessing `orientation` directly on the class and below you do it on `currentTourStep`. If I understand correctly it should be only accessed with the first approach.

lovaiible replied:
You are right, I missed that part. I changed the code to `switch (this.currentTourStep.orientation)` now

maxr96 commented on src/main/webapp/app/guided-tour/guided-tour.component.ts:
```diff
@@ -173,68 +178,191 @@ export class GuidedTourComponent implements AfterViewInit, OnDestroy {
+    private isBottom(): boolean {
+        if (this.currentTourStep && this.currentTourStep.orientation) {
+            return (
+                this.currentTourStep.orientation === Orientation.BOTTOM ||
```
You should be able to use just `this.orientation` here since you already assign it in the beginning

lovaiible replied:
Yes, but it refers to the orientation of the step, therefore it should refer to `this.currentTourStep.orientation`

maxr96 commented on src/main/webapp/app/guided-tour/guided-tour.constants.ts:
```diff
@@ -38,8 +31,17 @@ export enum OverlayPosition {
 }
 
 export enum UserInteractionEvent {
+    // @ts-ignore
```
Why do you need `@ts-ignore` here?

lovaiible replied:
I removed it :)

madwau commented on src/main/webapp/app/guided-tour/guided-tour.component.ts:
```diff
@@ -173,68 +178,191 @@ export class GuidedTourComponent implements AfterViewInit, OnDestroy {
+                    }
                 }
             }
+            setTimeout(() => {
```
I assume this timeout is needed to let the previously triggered scrolling finish? Is it not possible to move the flip orientation before the scrolling without timeout since the scrolling should not affect whether the tour is in the viewport horizontally?

Currently, the 300ms are noticeable and the user can see the popup jumping into the flipped orientation. Not super serious but if this could be avoided, even better.

lovaiible replied:
I changed the logic a little bit and now it's possible without the delay of 300ms :)

lovaiible commented on src/main/webapp/app/guided-tour/guided-tour.component.ts:
```diff
@@ -173,68 +178,191 @@ export class GuidedTourComponent implements AfterViewInit, OnDestroy {
+
+                if (this.isBottom()) {
+                    elementInViewPort =
+                        top >= window.pageYOffset + this.topOfPageAdjustment + scrollAdjustment + stepScreenAdjustment && top + height <= window.innerHeight + window.pageYOffset;
```
Unfortunately this is not possible because of the typescript formatting guideline given by the prettier config

lovaiible commented on src/main/webapp/app/guided-tour/guided-tour.component.html:
```diff
@@ -7,7 +7,7 @@
         #tourStep
         *ngIf="currentTourStep"
-        class="tour-step tour-{{ currentTourStep.orientation }}"
+        class="tour-step tour-{{ this.orientation }}"
```
Changed that :)

lovaiible commented:
Thank you @maxr96 and @madwau for your review! I addressed your feedback and also fixed the issue for smaller screens (@madwau).  Could you have another look and approve my PR if everything works for you now?
